### DISCUSSION TITLE
completion box will flash when input character

### DISCUSSION THREAD

**Discussion by Eason0210** [joaotavora/eglot#807] [General]
Hi,

I use `eglot` with `company`, lsp server is `clangd/pyright`.
And I encouter an issue:

When I input new character, the completion list will update, but the completion box will flash while I input new character.

This will not happens when `eglot` is disconnected with LSP server.

I not sure this is the issue of company or eglot?

I tested on `emacs -Q `,and install` company `and `eglot`.

Any one known how to fixed it?

edit: this issue also happen on Arch linux (`sudo pacman -S emacs-git`)

Platform: macOS
Emacs version: 
```
In GNU Emacs 29.0.50 (build 1, x86_64-apple-darwin19.6.0, NS appkit-1894.60 Version 10.15.7 (Build 19H1615))
 of 2022-01-14 built on Mac-1642119124451.local
Repository revision: d476fd33ece51f506644cec737490eda96daf46c
Repository branch: HEAD
Windowing system distributor 'Apple', version 10.3.2113
System Description:  macOS 12.1
```

**Comment by skangas** [comment]
Could you explain in more detail what you mean by "flash" here? Thanks.

**Reply by Eason0210** [reply]
Thanks for your reply.

It looks like  the completion box shaking in a short time to refresh the candidate content and adjust the width and height of box.
Without eglot activated, company will works very smooth.

**Reply by skangas** [reply]
Based on your description, it is hard to say if the root cause for this issue is with company or eglot. Since the issue at least starts with the company display flashing, I would probably start by asking there, as they are in a better position to investigate that part. If it turns out that eglot is doing something wrong, you can then open a bug in our tracker and provide the necessary details. Does that make sense?

**Reply by Eason0210** [reply]
Thanks, It make sense.

**Reply by dgutov** [reply]
As long as the problem occurs only together with Eglot and not with any other backend, "asking there" might not help.

We probably need more info here (such as whether a non-default frontend was used for popup, such as `company-posframe` or `company-box`, and whether "flashing" depends on the speed with which you enter the next character).

But this sounds vaguely similar to https://github.com/joaotavora/sly/issues/468

**Reply by Eason0210** [reply]
Hi @dgutov,

Thanks for your information.
The behavior looks similar to the video shows in [joaotavora/sly#468](https://github.com/joaotavora/sly/issues/468), but it is nothing to do with speed I input.
I used the default setting for company and eglot.

Steps to reproduce the issue:
1. start emacs with `emacs -Q`
2. install company and eglot
 - M-x package-install, RET, company
 - M-x package-install, RET, eglot
3. create a C++ file( you can use any language you like, I just tested `clangd `, `pyright` and `gopls`)
   M-x find-file test.cpp
4. enable company-mode and eglot
 - M-x company-mode
 - M-x eglot
5. input codes below, you will see the candidate box will "flashing" when you input character

```c++
#include <iostream>

int main ()
{
    std::cout << "Hello" << '\n';
    //   ^
    // when you input the std::co, you will see the flashing
    return 0;
}
```
6. M-x eglot-shutdown, and try the snippet again, you can see the company box is normal.

**Reply by dgutov** [reply]
So the problem is still there even if you try to type *v e r y    s l o w l y*?

**Reply by Eason0210** [reply]
> So the problem is still there even if you try to type _v e r y s l o w l y_?

Yes

**Comment by Eason0210** [comment]
@dgutov @skangas 

You can see the vedio below:

https://user-images.githubusercontent.com/47546452/149644665-5ee41b45-56cd-4d9b-8076-30af910d8292.mov

**Reply by dgutov** [reply]
Thanks for the video.

Not quite sure, but it seems like the arrival of the diagnostic causes extra redisplay..?

**Reply by Eason0210** [reply]
> Not quite sure, but it seems like the arrival of the diagnostic causes extra redisplay..?

@dgutov May be, and I can see there are a number showing in mode-line , such as [eglot:src/3].
Unfortunately, this info not showed in the above video, I record a new one that included it.

so, I wanted to known how to disabled showing the number in mode-line, and then may be the issue will be fixed?
**_update: I disabled the  mode-line info, but the issue still existed_**

https://user-images.githubusercontent.com/47546452/149646029-1e94b742-bd66-47ba-87db-69cac428a230.mov

**Reply by dgutov** [reply]
I have tried reproducing this, but the latest version of Eglot is even more blinky over here: it basically blinks the popup after every input. Even with `emacs -Q`.

The previous version (one I had before the update: 20210611.2249) didn't look like that.

I also couldn't disable `flymake-mode` -- the mode toggled off, but the diagnostics kept being highlighted and refreshed!

I don't have the time to dig into this, someone with better knowledge of Eglot should do it.

**Reply by Eason0210** [reply]
Hi @dgutov , @skangas 

@dgutov  you are right, I can confirm it.

 I tested with latest company [c25f1fb](https://github.com/company-mode/company-mode/commit/c25f1fbc3850e36e6521b77fa1641d5583365d8b) and  tried to use magit to biset `eglot,` and find that eglot [670ee](https://github.com/joaotavora/eglot/commit/670ee61169419e8e82b0bb115338bd5abbc44640) is the first bad commit. 
Before that commit, it works very well with company mode.

May be this information can help you to find the bug.
```
670ee61169419e8e82b0bb115338bd5abbc44640 is the first bad commit
```

**Reply by joaotavora** [reply]
Thanks. That commit seems legit, it's using the existing `completion-at-point-functions`  API. Apparently the contributor that sent it in and tested it is not using company as the frontend.  And from a cursory reading seems like these problems also don't apply to other frontends.

**Reply by dgutov** [reply]
I'm not 100% sure why it leads to blinking (let me know of your further findings). Probably there is an extra redisplay triggered by the `prefix` handler. Child frame based frontends are obviously immune to that.

But `:exclusive 'no` is a mess:

>                ;; FIXME: Here we'd need to decide whether there are
>               ;; valid completions against the current text.  But this depends
>               ;; on the actual completion UI (e.g. with the default completion
>               ;; it depends on completion-style) ;-(
>               ;; We approximate this result by checking whether prefix
>               ;; completion might work, which means that non-prefix completion
>               ;; will not work (or not right) for completion functions that
>               ;; are non-exclusive.

And yes, it does break non-prefix matching (when there are no proper prefix matches in the current set).

**Reply by joaotavora** [reply]
Seems to work here with other non-company frontends and for the author of that commit.  I haven't tested personally. :shrug:

**Reply by dgutov** [reply]
They just didn't notice the regression.

**Reply by joaotavora** [reply]
I don't know what you're referring to. If you have noted some regression others haven't, use the [appropriate form](https://github.com/joaotavora/eglot/issues/new?assignees=&labels=&template=bug_report.md&title=) to communicate it.

**Reply by dgutov** [reply]
https://github.com/joaotavora/eglot/issues/812

**Comment by Eason0210** [comment]
@skangas  @dgutov 

I tried to install only  eglot  and[ corfu](https://github.com/minad/corfu), the  completion box will not flashing.
NOTE: Corfu uses child frames to show the popup. 


https://user-images.githubusercontent.com/47546452/149651965-d1fd5a31-c01b-45b8-bb36-f1897530a892.mov

**Reply by dgutov** [reply]
> I tried to install only eglot and corfu, the completion box will not flashing.
NOTE: Corfu uses child frames to show the popup.

FWIW, if you try `company-posframe` (an alternative frontend to company), you should have the same result. It also uses child frames.

**Reply by Eason0210** [reply]
> if you try `company-posframe` (an alternative frontend to company)

Thanks, good to known that.

**Reply by jdtsmith** [reply]
`:exclusive no` causes very aggressive try-completion on every key press. Corfu has specific patch code to work around this. See https://github.com/minad/corfu/issues/9#issuecomment-960419376

**Reply by joaotavora** [reply]
Interesting, anyway, we're probably going to revert `:exclusive no`.  It causes functional problems problems even without even worrying about performance or "aggressiveness". As described in #812
